lib/Conversion/NVGPUToLLVM/NVGPUToLLVMPass.cpp

```diff
@@ -1105,43 +980,71 @@ class OffsetOfStmatrixV4OpPattern : public mlir::RewritePattern {
       } else if (rowStride == 16) {
         perPhase = 4;
         maxPhase = 2;
+      } else {
+        assert(false && "Unsupported rowStride");
       }
 
-      Value iterOfCol = udiv(elemIdx, i32_val(8));
-      Value myRow = add(rowOfWarp, and_(threadId, i32_val(0xf)));
-      Value myCol =
-          mul(and_(lshr(threadId, i32_val(4)), i32_val(0x1)), i32_val(8));
-      myCol = add(myCol, mul(iterOfCol, i32_val(16)));
-
-      Value offset0 =
-          mul(udiv(myCol, i32_val(rowStride)), i32_val(leadingDimOffset));
-      myCol = urem(myCol, i32_val(rowStride));
-
-      Value phase = urem(udiv(myRow, i32_val(perPhase)), i32_val(maxPhase));
-
-      Value lineOffset =
-          add(mul(urem(myRow, i32_val(perPhase)), i32_val(rowStride)), myCol);
-      Value colOffset =
-          add(mul(xor_(udiv(lineOffset, i32_val(8)), phase), i32_val(8)),
-              urem(lineOffset, i32_val(8)));
-      Value offset1 =
-          add(mul(udiv(myRow, i32_val(perPhase)), i32_val(64)), colOffset);
-
-      Value res = add(offset1, offset0);
-
-      rewriter.replaceOp(op, {res});
+      ptxAsm =
```

Comment:
why do we want to generate inline asm for this one and the one above? Is it for readability? Any concerns that we may be passing on potential llvm optimizations?

Reply:
Good point. I think it's mostly for readability and maintainability. We shouldn't compromise on llvm optimizations though. @ptillet any thoughts on this ?

Reply:
Is it possible to get the post-llvm optimized code and make it inline ptx ?

Reply:
> Is it possible to get the post-llvm optimized code and make it inline ptx ?

I guess but it still misses potential optimizations with the code around this function (thing like constant folding, cse, etc...). I don't know if there would be any opportunity there as I haven't looked enough in details how this is used to have a feeling

Reply:
Yes, this is for maintainability. I am generously assuming that ptxas will do any optimization applicable.

Reply:
In the grand scheme of things it is also not clear to me what this op is doing in this dialect. It is mostly just math